gen/field.go

```diff
@@ -95,22 +104,79 @@ func (f fieldGroupGenerator) DefineStruct(g Generator) error {
 		}`,
 		f,
 		TemplateFunc("tag", func(f *compile.FieldSpec) string {
-			// We want to add omitempty if the field is an optional struct or
-			// primitive to redure "null" noise. We won't add omitempty for
-			// optional collections because omitempty doesn't differentiate
-			// between nil and empty collections.
-
-			if (isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.Required {
-				return fmt.Sprintf("`json:\"%s,omitempty\"`", f.Name)
-			}
-
-			return fmt.Sprintf("`json:%q`", f.Name)
-			// TODO(abg): Take go.tag and js.name annotations into account
+			return generateTags(f)
 		}),
 		TemplateFunc("declFieldName", f.declFieldName),
 	)
 }
 
+// generateTags parses the annotation on the thrift field and creates the resulting go tag
+func generateTags(f *compile.FieldSpec) string {
+	// start with default json tag
+	tags := []string{
+		compileTag(jsonTagName, compileJSONTagValue(f, f.Name)),
+	}
+
+	// process go tags and overwrite json tag if specified in thrift annotation
+	if goAnnotation := f.ThriftAnnotations()[goTagName]; goAnnotation != "" {
+		if jt := reflect.StructTag(goAnnotation).Get(jsonTagName); jt != "" {
+			// truncate current tags to remove current json tag and compile the one defined in thrift
+			goTags := annotationsFromRawTag(goAnnotation)
+			goTags[jsonTagName] = compileJSONTagValue(f, jt)
+			tags, f.Annotations[goTagName] = tags[1:], annotationsToRawTag(goTags)
+		}
+		tags = append(tags, strings.Split(f.Annotations[goTagName], goTagDelimiter)...)
+	}
+	// TODO (abg): Take js.name annotation into account
+	return fmt.Sprintf("`%s`", strings.Join(tags, goTagDelimiter))
+}
+
+// annotationsFromRawTag constructs an annotations map from a raw tag string
+// `json:"foo,required" bson:"bar"` -> {"json": "foo,required", "bson": "bar"}
+func annotationsFromRawTag(a string) compile.Annotations {
+	ts := map[string]string{}
+	for _, t := range strings.Split(a, goTagDelimiter) {
+		k := strings.Split(t, goTagSep)[0]
+		ts[k] = reflect.StructTag(t).Get(k)
+	}
+	return ts
+}
+
+// annotationsToRawTag constructs a raw tag string from an annotations map
+// {"json": "foo,required", "bson": "bar"} -> `json:"foo,required" bson:"bar"`
+func annotationsToRawTag(t compile.Annotations) string {
+	var a []string
+	for k, v := range t {
+		a = append(a, compileTag(k, v))
+	}
+	return strings.Join(a, goTagDelimiter)
+}
+
+// compileTag constructs a single go tag from the tag name and value
+func compileTag(k, v string) string {
```

Comment:
Naming: This isn't really compilation. You're building a string representation of the tag from the key-value pair.

You don't really need this function. You can inline the logic in annotationsToRawTag. The other usage of this becomes unnecessary once you switch to `map[string]string` up there.